pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,223 @@
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+type singletonState struct {
+	sync.RWMutex
+	wObjSync map[util.ObjectIdentifier]bool
+}
+
+func (ss *singletonState) get(id util.ObjectIdentifier) bool {
+	ss.Lock()
+	defer ss.Unlock()
+
+	return ss.wObjSync[id]
+}
+
+func (ss *singletonState) set(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	ss.wObjSync[id] = true
+}
+
+func (ss *singletonState) unset(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	ss.wObjSync[id] = false
+}
+
+func (ss *singletonState) addIfNotExist(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	if _, ok := ss.wObjSync[id]; !ok {
+		ss.wObjSync[id] = false
+	}
+}
+
+func (ss *singletonState) delete(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	delete(ss.wObjSync, id)
+}
+
+func (c *Controller) buildSingletonStateAndOptionallyReconcile(ctx context.Context, reconcile bool) error {
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Building the desired state for singleton statuses")
+
+	for wObjIdentifier := range c.singletonState.wObjSync {
+		gvr := schema.GroupVersionResource{Group: wObjIdentifier.GVK.Group, Version: wObjIdentifier.GVK.Version, Resource: wObjIdentifier.Resource}
+		lister, found := c.listers.Get(gvr)
+		if !found {
+			return fmt.Errorf("could not get lister for gvr: %s", gvr)
+		}
+		_, err := lister.ByNamespace(wObjIdentifier.ObjectName.Namespace).Get(wObjIdentifier.ObjectName.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				c.singletonState.delete(wObjIdentifier)
+			}
+			return err
+		}
+	}
+
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, bdg := range allBdgs.Items {
+		logger.V(2).Info("Got Binding while building singleton state", "name", bdg.Name)
+		for _, item := range bdg.Spec.Workload.NamespaceScope {
+			nsWObjRef := item.NamespaceScopeDownsyncObject
+			lister, found := c.listers.Get(schema.GroupVersionResource(nsWObjRef.GroupVersionResource))
+			if !found {
+				return fmt.Errorf("could not get lister for gvr: %s", nsWObjRef.GroupVersionResource)
+			}
+			nsWObj, err := lister.ByNamespace(nsWObjRef.Namespace).Get(nsWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := nsWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: nsWObjRef.Group, Version: nsWObjRef.Version, Kind: nsWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   nsWObjRef.Resource,
+					ObjectName: cache.NewObjectName(nsWObjRef.Namespace, nsWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should have status synced", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName)
```

Comment:
This statement is going to be executed many times for the same object, right? I am not a fan of spamming the debug log at `-v=2` with such low level boring stuff.

Reply:
I totally agree. I temporarily (in this _WIP_ PR) put a 2 here just because I wanted to see it when I was debugging with `-v 2`. Say I put 4s and do `-v 4` then the logging will be too verbose and the messages will be hard to find.

Reply:
Now all the log levels are regularized by following the content of #2390